### PR TITLE
make deploy-group deletion not be 1-off controller logic, but use mod…

### PR DESCRIPTION
**Pull request by grosser**
…el validations

 - controllers should only use "save/destroy/soft_delete" and not know about model validations
 - models should know when they are valid / when they are allowed to be deleted

 - fixed that we should not destroy dependencies before soft_delete hook is started (SoftDeleteWithDestroy bug) 
 - fixed we should not set a permalink before validations are done so redirects do not break and forms do not show weird permalinks

screen shot 2019-03-07 at 1 44 40 pm

### CODE REVIEW

File: test/controllers/deploy_groups_controller_test.rb [hunk 1]
```diff
@@ -75,10 +75,15 @@
 
       it "compares environment" do
         var = EnvironmentVariable.create!(scope: deploy_groups(:pod2), parent: stage.project, name: "Foo", value: "bar")
+        group = EnvironmentVariableGroup.create!(name: "G1")
+        var2 = EnvironmentVariable.create!(scope: deploy_groups(:pod2), parent: group, name: "Bar", value: "baz")
```
Comment (grosser) [inline test/controllers/deploy_groups_controller_test.rb]: unrelated cleanup since this had missing coveragge

**Comment by grosser** [comment]
... now  it still runs validations when we use `validate: false`, so will need a workaround for that ...

**Review by sathishavm — APPROVED** [review]
LGTM
